ENH: sparse: add CSR/CSC `_broadcast_to` method

anushkasuyal opened the pull request:
This PR introduces broadcasting support for sparse CSR 2-D arrays (mentioned in #2128, #7442). I am currently working on extending support for COO sparse arrays to n-dimensions (see #21169 and #21197), where element-wise operations for n-D arrays will allow broadcasting (aligning with and replicating NumPy's array operation behaviours).

The enhancement implemented by this PR would enable broadcasting for element-wise operations on 2-D sparse arrays (which are routed via CSR), which becomes essential for ensuring consistent behaviour between 2-D and n-D COO arrays, and also across sparse arrays of all formats.

anushkasuyal committed:
broadcast_to added for CSR matrices

anushkasuyal committed:
correction made in broadcast_to, tests added

anushkasuyal committed:
eliminate_zeros removed

anushkasuyal committed:
.copy() added

anushkasuyal committed:
added NotImplementedError for CSC

anushkasuyal committed:
lint error fixed

perimosocordiae commented on scipy/sparse/_compressed.py:
```diff
@@ -1407,6 +1407,48 @@ def _divide_sparse(self, other):
 
 
+    def broadcast_to(self, shape):
+        if self.format != 'csr':
```
Let's move this method to the CSR class directly, which will avoid the need to check format here.

anushkasuyal replied:
Moved to class `_csr_base` - fd933de72b8289902921c5b5606e1ad980c8751e

perimosocordiae commented on scipy/sparse/_compressed.py:
```diff
@@ -1407,6 +1407,48 @@ def _divide_sparse(self, other):
+        old_shape = self.shape
+        
+        if old_shape == shape:
+            return self.copy()
```
`np.broadcast_to()` returns a view, not a copy. In our case, maybe we could pass `copy=False` as a keyword argument, similar to a few of the other sparse methods.

anushkasuyal replied:
Would that just be something like this - https://github.com/scipy/scipy/pull/21412/commits/a1811faaee68c6de5f2beac6c5a038e2e91dd402?

dschult replied:
That is close, but it doesn't return a copy when `copy is True`.  How about something like this?
```python
return self.copy() if copy else self
```

anushkasuyal replied:
Addressed at e67d2f7e9776c823400e58a5e3a52ead69efbb25.

perimosocordiae commented on scipy/sparse/_compressed.py:
```diff
@@ -1407,6 +1407,48 @@ def _divide_sparse(self, other):
+            raise ValueError("Target shape must be a tuple of length 2.")
+
+        # Ensure the old shape can be broadcast to the new shape
+        if any((o != 1 and o != n) for o, n in zip(old_shape, shape)):
```
Can we use `np.broadcast_shapes(old_shape, shape)` here?

anushkasuyal replied:
Changed at https://github.com/scipy/scipy/pull/21412/commits/9cdbce7203f0e4bd3420ff27054a46dde5669c90.

perimosocordiae commented on scipy/sparse/_compressed.py:
```diff
@@ -1407,6 +1407,48 @@ def _divide_sparse(self, other):
+        
+        if old_shape[0] == 1 and old_shape[1] == 1:
+            # Broadcast a single element to the entire shape
+            data = np.full(shape[0] * shape[1], self.data[0])
```
Note: even for a 1x1 sparse array, we might have `len(self.data) > 1` due to non-canonical formats (i.e., duplicate entries).

I recommend calling `self.sum_duplicates()` after the `nnz == 0` check.

anushkasuyal replied:
Added at https://github.com/scipy/scipy/pull/21412/commits/188c598e178541096d39ee044ecc20b12581de2d.

anushkasuyal committed:
broadcast_to moved to _csr_base in _csr.py

anushkasuyal committed:
np.broadcast_shapes used

anushkasuyal committed:
add self.sum_duplicates()

anushkasuyal committed:
adding copy=False

dschult commented on scipy/sparse/tests/test_base.py:
```diff
@@ -4137,6 +4137,25 @@ def test_binop_explicit_zeros(self):
+        d = np.array([[7]])
+        e = np.array([[0]])
+        f = np.array([[0,0,0,0]])
+        res_a = csr_matrix(a).broadcast_to((2,3))
```
Some of these tests should check that a ValueError occurs when the shape doesn't match.  In my playing with it,  it looks like using `np.broadcast_shapes` isn't what we want. It checks if the shapes can be broadcast to a unified shape. We need some tests to catch that.  Perhaps something like:
```python
    with pytest.raises(ValueError, match="cannot be broadcast"):
        csr_matrix([[1, 2, 3], [3, 2, 1]], (2, 1))
```
Note that `np.broadcast_shape((2, 3), (2, 1))` returns `(2, 3)` without raising. So I think your original check is needed.

anushkasuyal replied:
Ah right, got it. I've reverted the change and added tests to ensure that a ValueError is raised (96a3c54910b563bba252ca0c728af2295c8ad7ff).

anushkasuyal committed:
copy argument passed in broadcast_to

anushkasuyal committed:
reverted ValueError condition, modified tests

dschult commented:
This PR introduces a new method `csr.broadcast_to` but it doesn't use it anywhere. And it doesn't introduce `coo.broadcast_to` which the later PRs do have in their diffs..., so:
Is this PR meant to fix the PRs mentioned in the OP as well as be available for nD -- or is that for a separate PR?
and why doesn't this PR include the addition of `coo.broadcast_to`?

Also, the `numpy` version of `broadcast_to` does not copy any data. Instead, it adjusts the stride of the array so the same data gets reused with new indexes.  Have you tried to think about how broadcasting could work for sparse arrays without copying the data?  (That's almost certainly a different PR. Just wondering if you looked into that.)

anushkasuyal commented:
> This PR introduces a new method `csr.broadcast_to` but it doesn't use it anywhere. And it doesn't introduce `coo.broadcast_to` which the later PRs do have in their diffs..., so: Is this PR meant to fix the PRs mentioned in the OP as well as be available for nD -- or is that for a separate PR? and why doesn't this PR include the addition of `coo.broadcast_to`?

I made a separate PR for CSR broadcasting as I noticed a few comments in `_compressed.py` which labelled it as TODO. As huge PRs with a large amount of methods are harder to review, I thought it'd be better to have CSR broadcasting in a separate PR of its own, so that if in case the third PR doesn't for some reason get merged, atleast the CSR broadcasting is still up for anyone to use!

This PR only defines the method right now. In the third big PR that I will be making (following PR1 with the basic methods and PR2 with matmul, dot etc.), we introduce broadcasting for element-wise operations such as add, sub, multiply etc., which is currently not present for sparse arrays I think. While I was adding broadcasting for COO operations, I realised that it will also need to be incorporated for CSR and the other formats, because we can't have inconsistencies in whether we want our arrays to allow broadcasting or not. I think most formats route their element-wise operations via CSR, which is why I then added broadcasting for CSR.

Not adding COO broadcasting to this PR didn't have a specific reason, I just added it to the matmul etc. PR because I introduced it before working on those methods. If you prefer I have it in this PR, I can definitely make that addition in a commit!

> Also, the `numpy` version of `broadcast_to` does not copy any data. Instead, it adjusts the stride of the array so the same data gets reused with new indexes. Have you tried to think about how broadcasting could work for sparse arrays without copying the data? (That's almost certainly a different PR. Just wondering if you looked into that.)

I haven't had a think about implementing it without copying the data, but I will definitely look into that.

dschult commented:
> I made a separate PR for CSR broadcasting as I noticed a few comments in _compressed.py which labelled it as TODO. As huge PRs with a large amount of methods are harder to review, I thought it'd be better to have CSR broadcasting in a separate PR of its own, so that if in case the third PR doesn't for some reason get merged, atleast the CSR broadcasting is still up for anyone to use!

I interpret those `TODO` comments as we want to implement the feature of correctly broadcasting those results when the inputs are sparse. That is different from creating a user-facing function to do the broadcasting and return the sparse result.  I'm a little worried about this function being user-facing because we don't usually want to actually copy the rows/columns to do the broadcasting. In NumPy we change the "stride" so that it provides a view that mimics the bigger array. I could imagine this function expanding the sparse representation to fill all memory. Can we find a way to do this without copying all the other coords and data?  If not I guess I'm OK with providing this function if others see a need for it, but we shouldn't use this approach inside our functions when we do broadcasting to compute a product or binopt. I think there must be better ways.

Thanks for this PR being separate from the big one. I'll take a look there now to see how broadcasting is handled. :)

anushkasuyal commented:
> Can we find a way to do this without copying all the other coords and data? 

The COO `broadcast_to` in the other bigger PR also works by copying the data. I did give extensive thought to if we could follow a "view" approach for COO arrays too like we do for dense ones in numpy, but I'm unable to think of a way to do this in sparse because I'd assume we can't simply adjust the strides like we do for dense arrays (which resemble a grid, unlike sparse arrays). Same with CSR `broadcast_to`. I think the main reason to add it was to ensure that COO and CSR didn't perform differently (i.e. that one broadcasted and one didn't), but I realise that comes at the cost of physically expanding `coords` and `data`.

dschult commented:
Right... I've looked into this a bunch now too and the closest I can come to copy-free `tile` functionality is np.broadcast to a 2D array and then use `.flat` to make a flatiter instance. But that doesn't work nicely with the C++ routines. So, I'm changing my perspective. It seems like copying the index values is what is needed to get broadcasting to work with sparsetools here.

I'm still uncomfortable making this a public-facing function. Perhaps there is a use-case I'm not seeing. What do you think of adding this method as a private helper method `_broadcast_to`? Or, can you see a use-case for users wanting to broadcast (with copy)?  

Perhaps @perimosocordiae can weigh in on the choice of a public/private method. He seemed to be OK with public so long as there is a `copy=False` keyword arg. But I believe it will be a copy unless no broadcasting is done.

anushkasuyal commented:
> I'm still uncomfortable making this a public-facing function. Perhaps there is a use-case I'm not seeing. What do you think of adding this method as a private helper method `_broadcast_to`? Or, can you see a use-case for users wanting to broadcast (with copy)?

That makes complete sense to me! I just made it a public-facing function because `np.broadcast_to` is one, there was no other reasoning behind it. Should I change this one (and the `COO` one in the other PR) to `_broadcast_to`?

dschult commented:
Making it a private "helper" method would be my preference. It's easier to make it public later than to make it private later. :)

anushkasuyal committed:
_broadcast_to made helper function

anushkasuyal commented:
Done at 71165c888b01b9feb9599818153b29f7bb9bc72a. Is it okay to still have it inside the class or do we want to change the `self` argument name and have it as a function outside the class?

perimosocordiae commented:
It looks like something went wrong with the latest merge, and now there are a bunch of extra commits in the PR. Hopefully it's not too hard to rebase the branch to fix things up.

perimosocordiae commented:
+1 to making this a private helper function.

Keeping it as a free function instead of a method also sounds good.

dschult commented:
I think something close to the following will get this PR back in shape without conflicts.
I can push it up if you like.
```bash
git checkout main
git pull upstream main  # you might have named the scipy repo something other than upstream
git checkout add-csr-broadcasting

git checkout -b add-csr-broadcasting-backup   # creates an easy backup in case something goes bad
git checkout add-csr-broadcasting

git reset --hard 96a3c54            # this backs out the previous changes
git cherry-pick 71165c888b0       # this applies the one commit that you made after the rebase/merge
git rebase -i main   # opens an editor that should have about 12 commits listed. no changes needed just save/quit
git submodule update --init --recursive
```
The rebase command should bring in the lastest version of main without any conflicts. At least it did on my setup. I also did `git submodule update --init --recursive` and `git status` and `git log -3` a bunch of times to check where I was in the process. But this is the basic process. 

I think what went wrong last time was that the `main` that you pulled in was not updated to the latest scipy repo `main` branch. But I'm not sure.

I can push this up to your branch if want. Don't spend lots of time on it.

anushkasuyal commented:
Thanks @dschult for the steps and @perimosocordiae ! I've fixed the branch now. :)

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,47 @@ def tobsr(self, blocksize=None, copy=True):
+        if old_shape == shape:
+            return self.copy() if copy else self
+        
+        if len(shape) != 2:
```
Let's use the `sputil.check_shape()` function here, which also helpfully supports the 1d case if this is a csr_array.

anushkasuyal replied:
Done at a41da4b0720c0c1f44278fffa1ed133cbae9cb76.

dschult replied:
Does the newly added `check_shape` call need `allow_nd=self._allow_nd` parameter?

anushkasuyal replied:
Added it!

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,47 @@ def tobsr(self, blocksize=None, copy=True):
+        
+        if len(shape) != 2:
+            raise ValueError("Target shape must be a tuple of length 2.")
+
```
We should also check that the old_shape and new shape have the same length.

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,47 @@ def tobsr(self, blocksize=None, copy=True):
+                             f' broadcast to new shape {shape}')
+        
+        if self.nnz == 0: # array has no non zero elements
+            return self.__class__(np.zeros(shape, dtype=self.dtype), copy=False)
```
We can use the constructor that takes a shape directly, no need to create a dense array of zeros here.

anushkasuyal replied:
Done at a41da4b0720c0c1f44278fffa1ed133cbae9cb76.

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,47 @@ def tobsr(self, blocksize=None, copy=True):
+            return self.__class__(np.zeros(shape, dtype=self.dtype), copy=False)
+        
+        self.sum_duplicates()
+        if old_shape[0] == 1 and old_shape[1] == 1:
```
```suggestion
        if all(s == 1 for s in old_shape):
```

perimosocordiae commented on scipy/sparse/tests/test_base.py:
```diff
@@ -4137,6 +4137,31 @@ def test_binop_explicit_zeros(self):
+        d = np.array([[7]])
+        e = np.array([[0]])
+        f = np.array([[0,0,0,0]])
+        res_a = csr_matrix(a).broadcast_to((2,3))
```
Tests are failing because this method was renamed but the test wasn't updated to match.

anushkasuyal replied:
Fixed at 7cfdae829e38ef97b44922a03ca70822b5437801.

anushkasuyal committed:
method name updated in tests

anushkasuyal committed:
changes made to _broadcast_to

anushkasuyal committed:
Update scipy/sparse/_csr.py

Co-authored-by: CJ Carey <perimosocordiae@gmail.com>

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,51 @@ def tobsr(self, blocksize=None, copy=True):
+        if len(shape) != 2:
+            raise ValueError("Target shape must be a tuple of length 2.")
+
+        if len(old_shape) != len(shape):
+            raise ValueError("Original shape and target shape must be equal.")
```
This error message should say the lengths must be the same, not the shapes. :)

But can't we broadcast from a 1D array to a 2D array? Because CSR support 1D we should probably use some of the code from your COO `_broadcast_to`:
```python
       # Check if the new shape is compatible for broadcasting
        if len(new_shape) < len(old_shape):
            raise ValueError("New shape must have at least as many dimensions"
                             " as the current shape")

        # Add leading ones to shape to ensure same length as `new_shape` 
        shape = (1,) * (len(new_shape) - len(old_shape)) + tuple(old_shape)
```
and tailor it to the CSR case here.

anushkasuyal replied:
At the time of writing the method I think I only considered 2D to 2D cases. I've done some refactoring now which allows 1-D to 1-D (example (1,) to (4,)) and 1-D to 2-D, at https://github.com/scipy/scipy/pull/21412/commits/b510a8a4252ec521806f52a26fcd7e64720e2981.
One test in `test_base.py` is failing, which checks whether the dtype is `int32` (I think using `csr_array` might have caused this issue). Does this indicate that some implicit type conversion is occuring which I overlooked in my code, or is it just because of the structure of `csr_array`? If it's the second, should I test for 1-D to 2-D broadcasting in `test_csr.py` instead?

dschult replied:
The 1D tests should be in `test_csr.py` rather than `test_base.py`. 
You could move the whole test to `test_csr.py` to make life easier.

To keep the 2d parts in test_base.py we will need to:
- merge with main to get the new test_base changes
- change `csr_matrix` to `self.spcreator` so it will test with both csr_matrix and csr_array.
- put any part of the test that will be different when testing csr_matrix and csr_array inside an `if` clause like:
```python
    if not self.is_array_test:
        # do some test for the csr_matrix class 
    else:
        # do some test for the csr_array class
```
But again, it might be simpler to move it to `test_csr.py`

anushkasuyal replied:
I moved the test to `test_csr.py` for simplicity!

anushkasuyal committed:
_broadcast_to allows 1D to 1D/2D

anushkasuyal committed:
pull changes

anushkasuyal committed:
lint error fixed

anushkasuyal committed:
move test_broadcast_to to test_csr.py

anushkasuyal committed:
indentation fixed

dschult commented on scipy/sparse/tests/test_csr.py:
```diff
@@ -190,46 +190,46 @@ def test_mixed_index_dtype_int_indexing(cls):
+    res_d = csr_array(b)._broadcast_to((1,))
+    res_e = csr_array(e)._broadcast_to((4,0))
+
+    res_f = csr_matrix(f)._broadcast_to((2,4))
```
Is this line with `res_f` used in any tests?

anushkasuyal replied:
Nope, left it out by mistake, removed at https://github.com/scipy/scipy/pull/21412/commits/938fec8efcdc04d4489cfba1897d35368e5bbefb now!

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,70 @@ def tobsr(self, blocksize=None, copy=True):
+        
+        shape = check_shape(shape, allow_nd=(self._allow_nd))
+
+        both_are_1d = len(self.shape) == 1 and len(shape) == 1
+        if both_are_1d:
```
```suggestion
        if len(self.shape) == 1 and len(shape) == 1:
```

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,70 @@ def tobsr(self, blocksize=None, copy=True):
+            indptr = np.array([0, N])
+            return self.__class__((data, indices, indptr), shape=shape, copy=False)
+
+        if len(self.shape) == 1:
+            self = self.__class__(self.reshape(self._shape_as_2d))
+
+        old_shape = self.shape
```
No need to reshape self. We should just use `self._shape_as_2d`.
```suggestion
        # treat 1D as a 2D row
        old_shape = self._shape_as_2d
```

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,70 @@ def tobsr(self, blocksize=None, copy=True):
+        if len(shape) != 2:
+            raise ValueError('Target shape must be a tuple of length 2')
+
+        if len(old_shape) != len(shape):
+            raise ValueError('Original shape and target shape must have the'
+                             ' same length')
```
I think both the approach of reshaping 1D to 2D and the approach of using `_shape_as_2d` make this check strange. The only ways it would raise is if `shape` is >2D or both `shape` is 1D and self is not.
Perhaps we should combine this check with the if statement that checks broadcasting.

Actually, can we eliminate much of this checking code by using `np.broadcast_shapes`? Is that better? Or is this better?   :}

anushkasuyal replied:
Ah yes, this check doesn't really make sense anymore, especially since we check that `len(shape) == 2` in the line right before this. I've removed it at https://github.com/scipy/scipy/pull/21412/commits/8eed3d09249ade78045d1dc0da82825be3b19faa.
I think using `np.broadcast_shapes` will raise the issue you pointed out a few weeks ago - `(2,1)` and `(2,2)` are broadcastable, but we can't broadcast a `(2,2)` array into a `(2,1)` array.

dschult replied:
You are right that (1,2) and (2,2) are broadcastable. But I thought of the idiom that would check "broadcast_to" using "broadcast_shapes":
```python
    if np.broadcast_shapes(self.shape, shape) == shape:
```
But does it have any advantages over what we're doing in terms of reducing the checks?  We still need to check a bunch...  I was also hoping it might cut down on the number of exceptions to word/reword since we could use the numpy words. But that doesn't always work well either. :)

anushkasuyal replied:
I moved the check above to avoid checking for the shape being `(1,)` in the 1-D case. Does this make sense? - 60050c0277eb0f1e3e83d3838209a1755925b86b
If the shapes aren't broadcastable, `np.broadcast_shapes` will throw an error during the if statement. If they are broadcastable but in the wrong order, we throw the `"cannot be broadcast"` error.

anushkasuyal committed:
unused res_f removed

anushkasuyal committed:
shape len equality check removed

anushkasuyal committed:
Update scipy/sparse/_csr.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_csr.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
use np.broadcast_shapes

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,58 @@ def tobsr(self, blocksize=None, copy=True):
+        old_shape = self._shape_as_2d
+            
+        if len(shape) != 2:
+            raise ValueError('Target shape must be a tuple of length 2')
```
Since `shape` can be 1D for CSR, the wording of the message can become softer:
```suggestion
            raise ValueError(f'CSR broadcast_to cannot have shape >2D. Got {ndim}D')```

dschult reviewed (commented):
This looks ready to merge. 
I have two very minor suggestions below.

In the bigger picture:
This adds a `_broadcast_to` helper method of the CSR format class. 

One feature that is not addressed here is handling `idx_type`.

Future PRs can make this work for CSC (maybe in `_compressed.py` module).
Some of the ideas here can be used to update the COO format version when that gets merged.
Future PRs should use this helper to actually implement the motivating "Todo" comments and see how well it performs for those tasks.

Thanks Anushka for splitting this out as a separate PR.
@perimosocordiae can you have a look?

anushkasuyal committed:
Update scipy/sparse/_csr.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
ValueError modified

anushkasuyal committed:
blank line

anushkasuyal commented on scipy/sparse/tests/test_base.py:
```diff
@@ -4137,7 +4137,6 @@ def test_binop_explicit_zeros(self):
         assert a.multiply(b).nnz == 1
 
-
 TestCSR.init_class()
```
Done!

dschult commented on scipy/sparse/tests/test_csr.py:
```diff
@@ -188,3 +188,46 @@ def test_mixed_index_dtype_int_indexing(cls):
+    d = np.array([[7]])
+    e = np.array([[0]])
+    f = np.array([[0,0,0,0]])
+    res_a = csr_matrix(a)._broadcast_to((2,3))
+    res_b = csr_matrix(b)._broadcast_to((3,4))
+    res_c = csr_matrix(c)._broadcast_to((2,3))
+    res_d = csr_matrix(d)._broadcast_to((4,4))
+    res_e = csr_matrix(e)._broadcast_to((5,6))
+    res_f = csr_matrix(f)._broadcast_to((2,4))
+    assert_array_equal(res_a.toarray(), np.broadcast_to(a, (2,3)))
+    assert_array_equal(res_b.toarray(), np.broadcast_to(b, (3,4)))
+    assert_array_equal(res_c.toarray(), c)
+    assert_array_equal(res_d.toarray(), np.broadcast_to(d, (4,4)))
+    assert_array_equal(res_e.toarray(), np.broadcast_to(e, (5,6)))
+    assert_array_equal(res_f.toarray(), np.broadcast_to(f, (2,4)))
```
```suggestion
    for csr_container in (csr_matrix, csr_array):
        res_a = csr_container(a)._broadcast_to((2,3))
        res_b = csr_container(b)._broadcast_to((3,4))
        res_c = csr_container(c)._broadcast_to((2,3))
        res_d = csr_container(d)._broadcast_to((4,4))
        res_e = csr_container(e)._broadcast_to((5,6))
        res_f = csr_container(f)._broadcast_to((2,4))
        assert_array_equal(res_a.toarray(), np.broadcast_to(a, (2,3)))
        assert_array_equal(res_b.toarray(), np.broadcast_to(b, (3,4)))
        assert_array_equal(res_c.toarray(), c)
        assert_array_equal(res_d.toarray(), np.broadcast_to(d, (4,4)))
        assert_array_equal(res_e.toarray(), np.broadcast_to(e, (5,6)))
        assert_array_equal(res_f.toarray(), np.broadcast_to(f, (2,4)))
```

dschult reviewed (commented):
I think the last (remove one blank line) commit from a suggestion only triggered the docs CI tests. So I'm going to commit this suggestion (which makes the tests cover both csr_matrix and csr_array for the 2D cases).

dschult committed:
update test to cover both csr_matrix and csr_array for 2D

anushkasuyal committed:
CSC broadcasting added, _broadcast_to moved to _compressed.py

anushkasuyal committed:
Merge branch 'add-csr-broadcasting' of https://github.com/anushkasuyal/scipy into add-csr-broadcasting

anushkasuyal committed:
lint error

rgommers commented on scipy/sparse/tests/test_csc.py:
```diff
@@ -96,3 +96,30 @@ def test_argmax_overflow(ax):
 
     assert A[ii, jj] == A[-2, -2]
+
+def test_broadcast_to():
```
This test code is mostly duplicate between `test_csc.py` and `test_csr.py`. I think the two can be merged fairly easily, and moved to `test_base.py`. Writing the iteration like:

```python
for container in (csr_matrix, csr_array, csc_matrix, csc_array):
```

Probably worth doing that before merging.

anushkasuyal committed:
tests moved

rgommers commented on scipy/sparse/tests/test_csc.py:
```diff
@@ -1,6 +1,6 @@
-from numpy.testing import assert_array_almost_equal, assert_
-from scipy.sparse import csr_matrix, csc_matrix, lil_matrix
+from numpy.testing import assert_array_almost_equal, assert_, assert_array_equal
+from scipy.sparse import csr_matrix, csc_matrix, lil_matrix, csc_array
```
these changes can be dropped now

anushkasuyal committed:
lint errors

anushkasuyal committed:
Merge branch 'main' into add-csr-broadcasting

anushkasuyal committed:
merge error

perimosocordiae reviewed (approved):
Thanks!

This current implementation involves a fair amount of data copying, which isn't necessarily what users would expect. My concern is that a broadcasted multiply may end up consuming all available RAM, when the numpy implementation would not require that.

That said, I'm happy to merge this as-is and we can track improvements to broadcasted sparse arrays in a separate issue.

stefanv reviewed (commented):
Thanks @anushkasuyal! Will also create a follow-up issue to track @perimosocordiae's comment.
